stdlib/runtime/numbers.gr

```diff
@@ -2807,414 +2804,376 @@ let rec expBySquaring = (y, x, n) => {
 provide let (**) = (base, power) => {
   let (==) = numberEq
   let (!=) = (x, y) => !numberEq(x, y)
+  let basePtr = WasmI32.fromGrain(base)
+  let powerPtr = WasmI32.fromGrain(power)
   if (base == 1 && power != 0) {
-    1
-  } else if (
-    isInteger(WasmI32.fromGrain(base)) && isInteger(WasmI32.fromGrain(power))
-  ) {
-    if (power < 0) expBySquaring(1, 1 / base, power * -1)
-    else expBySquaring(1, base, power)
+    return 1
+  } else if (isInteger(basePtr) && isInteger(powerPtr)) {
+    if (power < 0) return expBySquaring(1, 1 / base, power * -1)
+    else return expBySquaring(1, base, power)
+  } else if (isRational(basePtr) && isInteger(powerPtr)) {
+    // Apply expBySquaring to numerator and denominator
+    let numerator = WasmI32.toGrain(boxedRationalNumerator(basePtr)): Number
+    let denominator = WasmI32.toGrain(boxedRationalDenominator(basePtr)): Number
```

Comment:
`boxedRationalNumerator` and `boxedRationalDenominator` both just return raw bigint pointers, so these need to go through some transformation process to become proper numbers. You can use `reducedBigInteger`, though you'll need to `incRef` the numerator and denominator. It could be useful to have utilities for this. Could be good to check with @alex-snezhko to see if he added some when he did the rational work.

Reply:
`rationalNumerator` and `rationalDenominator` in `runtime/numbers` seem to be about what you want, though they specifically take the `Rational` type as an argument. I guess you could "fake" the type of the argument with something like `let baseRational = WasmI32.toGrain(WasmI32.fromGrain(base)): Rational` or maybe extract the stuff after the `WasmI32.fromGrain` calls from these functions [(link)](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/numbers.gr#L1151) into separate functions and reuse those function here

Reply:
Just chiming in that doing `WasmI32.toGrain(WasmI32.fromGrain(base))` is generally unsafe because it gives the GC an opportunity to collect your value. Even though you're making a new reference to it. We have #1149 though to make a utility to do this safely. (You can always add an incRef call, though.)

Reply:
But `rationalNumerator` and `rationalDenominator` sound really good 😛